awx/api/views/__init__.py

```diff
@@ -3824,6 +3824,82 @@ def get_queryset(self):
         return job.get_event_queryset().select_related('host').order_by('start_line')
 
 
+class JobJobEventsChildrenSummary(APIView):
+    '''
+    Special view to facilitate processing job output in the UI.
+    In order to collapse events and their children, the UI needs to know how
+    many children exist for a given event.
+    This view returns a JSON object where the key is the event counter, and the value
+    is the number of children (and grandchildren) events.
+    Only events with children are included in the output.
+
+    e.g. Demo Job Template job
+    given:
+        tuple(event counter, uuid, parent_uuid)
+        (1, '4598d19e-93b4-4e33-a0ae-b387a7348964', '')
+        (2, 'aae0d189-e3cb-102a-9f00-000000000006', '4598d19e-93b4-4e33-a0ae-b387a7348964')
+        (3, 'aae0d189-e3cb-102a-9f00-00000000000c', 'aae0d189-e3cb-102a-9f00-000000000006')
+        (4, 'f4194f14-e406-4124-8519-0fdb08b18f4b', 'aae0d189-e3cb-102a-9f00-00000000000c')
+        (5, '39f7ad99-dbf3-41e0-93f8-9999db4004f2', 'aae0d189-e3cb-102a-9f00-00000000000c')
+        (6, 'aae0d189-e3cb-102a-9f00-000000000008', 'aae0d189-e3cb-102a-9f00-000000000006')
+        (7, '39a49992-5ca4-4b6c-b178-e56d0b0333da', 'aae0d189-e3cb-102a-9f00-000000000008')
+        (8, '504f3b28-3ea8-4f6f-bd82-60cf8e807cc0', 'aae0d189-e3cb-102a-9f00-000000000008')
+        (9, 'a242be54-ebe6-4021-afab-f2878bff2e9f', '4598d19e-93b4-4e33-a0ae-b387a7348964')
+
+    output:
+        {
+        "1": 8,
+        "2": 6,
+        "3": 2,
+        "6": 2
```

Comment:
Even if this method can run in a reasonable amount of time for a high-output job, the _number of keys_ is going to make it difficult to return in the response. If there are a million events, will it even be able to _transmit_ the response before the uWSGI worker dies? As a whole-job endpoint, I find it doubtful that this endpoint will survive the test of real users.

Reply:
Then again, I get where you're coming from excluding events with _no_ children. Assuming the multi-million or multi-billion events are because of number of hosts... maybe. How do loops figure into this? If I run one host but loop a task a million times, what fraction of the events have children?

Reply:
yeah this is where things aren't clear me to me. The output here only includes events with children, which I am hoping is a small enough subset to reliably do this kind of thing. We may need to find a realistic, high output job to test this on.

Reply:
Running chatty_tasks with 1 host

```
In [6]: parents = set(event.parent_uuid for event in events)

In [7]: len(events)
Out[7]: 56

In [8]: len([event for event in events if event.uuid in parents])
Out[8]: 3
```

So seems fairly valid to cut out the bulk of data from loops.